src/mlpack/methods/reinforcement_learning/replay/episodic_replay.hpp

```diff
@@ -0,0 +1,271 @@
+  using ActionType = typename EnvironmentType::Action;
+
+  //! Convenient typedef for state.
+  using StateType = typename EnvironmentType::State;
+
+  /**
+  * Construct an instance of episode experience replay class.
+  * Everything initialized to 0
+  */
+  EpisodicReplay():
+      capacity(0),
+      position(0),
+      max_episode_len(0),
+      full(false)
+  { /* Nothing to do here. */ }
+
+  /**
+  * Construct an instance of episode experience replay class.
+  * @param capacity Total memory size in terms of number of episodes.
+  * @param dimension The dimension of an encoded state.
+  */
+  EpisodicReplay(const size_t capacity,
+                 const size_t max_episode_len,
+                 const size_t dimension = StateType::dimension) :
+      capacity(capacity),
+      position(0),
+      max_episode_len(max_episode_len),
+      full(false)
+  {
+    states.resize(capacity);
+    next_states.resize(capacity);
+    rewards.resize(capacity);
+    actions.resize(capacity);
+    isTerminal.resize(capacity); 
+  }
+
+
+  /**
+  * Store the given experience.
+  * @param state Given state.
+  * @param action Given action.
+  * @param reward Given reward.
+  * @param nextState Given next state.
+  * @param isEnd Whether next state is terminal state.
+  */
+  void Store(const StateType& state,
+             ActionType action,
+             double reward,
+             const StateType& nextState,
+             bool isEnd)
+  {
+    if(clear)
+    {
+      states[position].clear();
+      actions[position].clear();
+      rewards[position].clear();
+      next_states[position].clear();
+      isTerminal[position].clear();
+      clear = false;
+    }
+    states[position].push_back(state.Encode());
+    actions[position].push_back(action);
+    rewards[position].push_back(reward);
+    next_states[position].push_back(nextState.Encode());
+    isTerminal[position].push_back(isEnd);
+    if(isEnd||states[position].size()==max_episode_len)
+    {
+      position++;
+      clear = true;
+    }
+    if (position == capacity)
+    {
+      full = true;
+      position = 0;
+    }
+  }
+
+  /**
+  * Sample some episode.
+  * @param sampledStates Sampled encoded states.
+  * @param sampledActions Sampled actions.
+  * @param sampledRewards Sampled rewards.
+  * @param sampledNextStates Sampled encoded next states.
+  * @param isTerminal Indicate whether corresponding next state is terminal
+  *        state.
+  */
+
+  void Random_Episode(arma::mat& episodeStates,
+                      arma::icolvec& episodeActions,
+                      arma::colvec& episodeRewards,
+                      arma::mat& episodeNextStates,
+                      arma::icolvec& isTerminal)
+  {
+    size_t upperBound = full ? capacity : position;
+    int lo = 0;
+    int high = upperBound;
+    if(!upperBound)
+    {
+      high = 1;
+    }
+    int episodeNum = math::RandInt(lo,high);
+    std::vector<arma::colvec> temp = states[episodeNum];
```

Comment:
Looks like you could directly use `states[episodeNum]`? We could avoid the extra that way.